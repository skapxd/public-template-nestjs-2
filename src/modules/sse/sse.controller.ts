import { Controller, Get, Param, Post, Query, Req, Sse } from '@nestjs/common';
import { Cron, CronExpression } from '@nestjs/schedule';
import { ApiOkResponse, ApiProduces, ApiResponse } from '@nestjs/swagger';
import { Request } from 'express';

import { ApiResponseSSE } from '#/src/utils/decorators/api-response-sse.decorator';

import { CounterDto, NotificationsDto } from './sse.dto';
import { SseService } from './sse.service';

@Controller('sse')
export class SseController {
  constructor(private readonly sseService: SseService) {}

  @Get()
  render(@Query('id') id: string) {
    return this.sseService.render(id);
  }

  @Post()
  btn(@Query('counter') counter: number, @Query('id') id: string) {
    this.sseService.btn(counter, id);
  }

  @ApiOkResponse({ type: NotificationsDto, description: 'Notificaciones' })
  @Post('all')
  @Cron(CronExpression.EVERY_SECOND)
  cron() {
    this.sseService.cron();
  }

  @ApiProduces('text/stream')
  @ApiResponse({
    status: 200,
    type: CounterDto,
    description: 'Mensaje',
  })
  @ApiResponseSSE({
    event: NotificationsDto.event,
    type: NotificationsDto,
    description: 'Notificaciones',
  })
  @Sse(':id')
  sse(@Param('id') id: string, @Req() req: Request) {
    req.on('close', this.sseService.onCloseSse);

    return this.sseService.sse(id);
  }
}
