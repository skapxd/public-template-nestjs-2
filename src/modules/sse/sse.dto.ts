import { ApiProperty } from '@nestjs/swagger';

export class Counter {
  @ApiProperty({ default: 0 })
  counter: number;

  constructor(args: Counter) {
    Object.assign(this, args);
  }
}

export class CounterDto extends MessageEvent<Counter> {
  static event = 'counter';

  @ApiProperty({ default: CounterDto.event })
  event: string;

  @ApiProperty({ type: Counter })
  data: Counter;

  constructor(args: Counter) {
    super(CounterDto.event, { data: args });
  }
}

export class NotificationsDto extends MessageEvent<string> {
  static event = 'notification';

  @ApiProperty({ default: NotificationsDto.event })
  event: string;

  @ApiProperty({ default: 'notification value' })
  data: string;

  constructor(args: string) {
    super(NotificationsDto.event, { data: args });
  }
}
