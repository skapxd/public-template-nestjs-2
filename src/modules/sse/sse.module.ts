import { Module } from '@nestjs/common';
import { EventEmitterModule } from '@nestjs/event-emitter';
import { ScheduleModule } from '@nestjs/schedule';

import { SseController } from './sse.controller';
import { SseService } from './sse.service';

@Module({
  imports: [
    ScheduleModule.forRoot(),
    EventEmitterModule.forRoot({
      delimiter: '.',
      wildcard: true,
      global: true,
      verboseMemoryLeak: true,
    }),
  ],
  controllers: [SseController],
  providers: [SseService],
})
export class SseModule {}
