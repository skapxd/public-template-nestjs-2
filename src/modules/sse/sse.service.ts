import { randomUUID } from 'node:crypto';

import { Injectable, Logger } from '@nestjs/common';
import { EventEmitter2 } from '@nestjs/event-emitter';
import { fromEvent, map } from 'rxjs';

import { CounterDto } from './sse.dto';

@Injectable()
export class SseService {
  private readonly logger = new Logger(SseService.name);
  private counter = 0;

  constructor(private eventEmitter: EventEmitter2) {}

  async cron() {
    this.eventEmitter.emit('btn', { counter: 1, id: null });
  }

  render(uuid?: string) {
    const _uuid = uuid || randomUUID();

    return `
    <button onclick="fn()">this</button>
    <button onclick="fnAll()">all</button>

      <ul></ul>

      <script type="text/javascript">
      function fn() {
        fetch('sse?counter=1000&id=${_uuid}', {
        method: 'POST',
        });
      }

      function fnAll() {
        fetch('sse/all', {
          method: 'POST',
        });
      }

      const ul = document.querySelector('ul')
      const eventSource = new EventSource('/sse/${_uuid}');
      eventSource.onmessage = ({ data }) => {
        const message = document.createElement('li');
        message.innerText = 'New message: ' + data;
        ul.insertBefore(message, ul.firstChild);
      }
      </script>
    `.replace(/  /g, '');
  }

  btn(counter: number, id: string) {
    this.eventEmitter.emit('btn', { counter, id });
    this.logger.log(`${id} - ${counter}`);
    return { counter };
  }

  onCloseSse() {
    this.counter -= 1;
  }

  sse(id: string) {
    return fromEvent(this.eventEmitter, 'btn').pipe(
      map((_: any) => {
        this.counter += _.counter;

        if (_.id === id) {
          return new CounterDto({ counter: this.counter });
        }

        if (_.id === null) {
          return new CounterDto({ counter: this.counter });
        }
      }),
    );
  }
}
