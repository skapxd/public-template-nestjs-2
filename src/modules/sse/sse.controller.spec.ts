import { randomUUID } from 'node:crypto';

import { INestApplication } from '@nestjs/common';
import { Test, TestingModule } from '@nestjs/testing';
import EventSource from 'eventsource';
import request from 'supertest';

import { mainConfig } from '#/src/utils/main-config.no-spec';

import { SseModule } from './sse.module';

describe('SseController', () => {
  let app: INestApplication;

  beforeAll(async () => {
    const module: TestingModule = await Test.createTestingModule({
      imports: [SseModule],
    }).compile();

    app = module.createNestApplication();

    mainConfig(app);

    await app.init();

    await app.listen(0);
  });

  it('should be defined', () => {
    expect(app).toBeDefined();
  });

  it.skip('should recibe event', async () => {
    const uuid = randomUUID();
    const url = await app.getUrl();
    const eventSource = new EventSource(url + '/sse/' + uuid);

    const messagePromise = new Promise((resolve, reject) => {
      eventSource.addEventListener('counter', ({ data }) => {
        resolve(data);
      });

      eventSource.onerror = (err) => {
        reject(err);
      };
    });

    await request(app.getHttpServer())
      .post('/sse')
      .query({ counter: 1, id: uuid });

    const resp = await messagePromise;

    expect(resp).toMatchSnapshot();
  });

  it('should render html', async () => {
    const resp = await request(app.getHttpServer()).get('/sse?id=uuid');

    expect(resp.text).toMatchSnapshot();
  });
});
