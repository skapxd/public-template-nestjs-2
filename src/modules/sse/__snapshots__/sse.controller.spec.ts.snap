// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SseController > should recibe event 1`] = `"{"hello":"world","counter":1}"`;

exports[`SseController > should render html 1`] = `
"
<button onclick="fn()">this</button>
<button onclick="fnAll()">all</button>

<ul></ul>

<script type="text/javascript">
function fn() {
fetch('sse?counter=1000&id=uuid', {
method: 'POST',
});
}

function fnAll() {
fetch('sse/all', {
method: 'POST',
});
}

const ul = document.querySelector('ul')
const eventSource = new EventSource('/sse/uuid');
eventSource.onmessage = ({ data }) => {
const message = document.createElement('li');
message.innerText = 'New message: ' + data;
ul.insertBefore(message, ul.firstChild);
}
</script>
"
`;
